File: mobile/lib/core/money.dart
/// Money formatting that never goes through a floating-point number.
///
/// The API sends money as exact decimal strings ("1874.25"). Parsing those into
/// a Dart `double` would reintroduce binary floating-point error - precisely
/// what the backend's `Numeric(10,2)` columns exist to prevent. Everything here
/// operates on the string.
library;

const String rupee = '₹';

class _Parts {
  const _Parts(this.negative, this.whole, this.fraction);

  final bool negative;
  final String whole;
  final String fraction;
}

_Parts _split(String value) {
  final trimmed = value.trim();
  final negative = trimmed.startsWith('-');
  final unsigned = negative ? trimmed.substring(1) : trimmed;
  final dot = unsigned.indexOf('.');
  if (dot == -1) {
    return _Parts(negative, unsigned.isEmpty ? '0' : unsigned, '');
  }
  final whole = unsigned.substring(0, dot);
  return _Parts(negative, whole.isEmpty ? '0' : whole, unsigned.substring(dot + 1));
}

/// Group digits the Indian way: the last three, then pairs.
/// `1234567` becomes `12,34,567`.
String groupIndian(String digits) {
  if (digits.length <= 3) return digits;
  final last3 = digits.substring(digits.length - 3);
  final rest = digits.substring(0, digits.length - 3);

  final buffer = StringBuffer();
  for (var i = 0; i < rest.length; i++) {
    // A separator every two digits, counted from the right of `rest`.
    if (i > 0 && (rest.length - i) % 2 == 0) buffer.write(',');
    buffer.write(rest[i]);
  }
  return '$buffer,$last3';
}

/// Format a decimal string as `₹1,874.25`.
String formatMoney(String? value, {bool showSymbol = true}) {
  if (value == null || value.trim().isEmpty) {
    return showSymbol ? '${rupee}0.00' : '0.00';
  }
  final parts = _split(value);
  final paise = '${parts.fraction}00'.substring(0, 2);
  final body = '${groupIndian(parts.whole)}.$paise';
  return '${parts.negative ? '-' : ''}${showSymbol ? rupee : ''}$body';
}

/// Format a decimal string as whole rupees, e.g. `₹1,874`.
String formatMoneyShort(String? value) {
  if (value == null || value.trim().isEmpty) return '${rupee}0';
  final parts = _split(value);
  return '${parts.negative ? '-' : ''}$rupee${groupIndian(parts.whole)}';
}

/// Format `"25.0"` as `25% OFF`, dropping a pointless trailing `.0`.
String formatDiscount(String? value) {
  if (value == null || value.isEmpty) return '';
  final cleaned = value.endsWith('.0') ? value.substring(0, value.length - 2) : value;
  return '$cleaned% OFF';
}

/// True when the value represents a real saving rather than `"0.0"`.
bool hasDiscount(String? discountPercentage) {
  if (discountPercentage == null) return false;
  final parts = _split(discountPercentage);
  return parts.whole != '0' || parts.fraction.replaceAll('0', '').isNotEmpty;
}

/// Compare two decimal strings exactly, without floating point.
int compareMoney(String a, String b) {
  final pa = _split(a);
  final pb = _split(b);
  if (pa.negative != pb.negative) return pa.negative ? -1 : 1;

  BigInt scaled(_Parts p) => BigInt.parse('${p.whole}${'${p.fraction}00'.substring(0, 2)}');
  final diff = scaled(pa) - scaled(pb);
  final magnitude = diff == BigInt.zero ? 0 : (diff > BigInt.zero ? 1 : -1);
  return pa.negative ? -magnitude : magnitude;
}

File: desktop/src/lib/money.ts
/**
 * Money formatting that never converts through a float.
 *
 * The API sends money as exact decimal strings ("1874.25"). Parsing those into
 * a JS `number` would reintroduce binary floating-point error - the very thing
 * the backend's Numeric(10,2) columns exist to prevent. Everything here works
 * on the string.
 */

const INR = '₹';

/** Split "1874.25" into its sign, integer and fractional parts. */
function parts(value: string): { negative: boolean; whole: string; fraction: string } {
  const trimmed = (value ?? '').trim();
  const negative = trimmed.startsWith('-');
  const unsigned = negative ? trimmed.slice(1) : trimmed;
  const [whole = '0', fraction = ''] = unsigned.split('.');
  return { negative, whole: whole || '0', fraction };
}

/**
 * Group digits the Indian way: the last three, then pairs.
 * 1234567 -> "12,34,567"
 */
function groupIndian(digits: string): string {
  if (digits.length <= 3) return digits;
  const last3 = digits.slice(-3);
  const rest = digits.slice(0, -3);
  return `${rest.replace(/\B(?=(\d{2})+(?!\d))/g, ',')},${last3}`;
}

/** Format a decimal string as "₹1,874.25". */
export function formatMoney(value: string | null | undefined, showSymbol = true): string {
  if (value === null || value === undefined || value === '') return showSymbol ? `${INR}0.00` : '0.00';
  const { negative, whole, fraction } = parts(value);
  const paise = (fraction + '00').slice(0, 2);
  const body = `${groupIndian(whole)}.${paise}`;
  return `${negative ? '-' : ''}${showSymbol ? INR : ''}${body}`;
}

/** Format a decimal string as a compact whole-rupee figure, e.g. "₹1,874". */
export function formatMoneyShort(value: string | null | undefined): string {
  if (!value) return `${INR}0`;
  const { negative, whole } = parts(value);
  return `${negative ? '-' : ''}${INR}${groupIndian(whole)}`;
}

/** Format "25.0" as "25%", dropping a pointless trailing ".0". */
export function formatPercent(value: string | null | undefined): string {
  if (!value) return '0%';
  const cleaned = value.replace(/\.0$/, '');
  return `${cleaned}%`;
}

/** Group a plain integer count, e.g. 12345 -> "12,345". */
export function formatCount(value: number): string {
  return groupIndian(String(Math.trunc(Math.abs(value)))).replace(/^/, value < 0 ? '-' : '');
}

/**
 * Compare two decimal strings without floating point.
 * Returns a negative number, zero, or a positive number like a comparator.
 */
export function compareMoney(a: string, b: string): number {
  const pa = parts(a);
  const pb = parts(b);
  if (pa.negative !== pb.negative) return pa.negative ? -1 : 1;
  const scale = (p: { whole: string; fraction: string }) =>
    BigInt(p.whole + (p.fraction + '00').slice(0, 2));
  const diff = scale(pa) - scale(pb);
  const magnitude = diff === 0n ? 0 : diff > 0n ? 1 : -1;
  return pa.negative ? -magnitude : magnitude;
}

/** True when the string is a well-formed money amount with at most 2 decimals. */
export function isValidMoney(value: string): boolean {
  return /^\d{1,8}(\.\d{1,2})?$/.test(value.trim());
}
